#!/bin/bash

##to debug
#set -e
#set -v
#set -x

########################################
## A bash script to run the gplas pipeline
## This script has been converted and transformed from the script present in the gitlab repo 'bactofidia' by aschuerch

while getopts ":i:n:s:c:t:x:r:f:h" opt; do
 case $opt in
   h)
   cat figures/logo.txt
   echo -e "Welcome to the user guide of gplas:\n"
   echo -e "Basic usage example: ./gplas.sh -i mygraph.gfa -c mlplasmids -s 'Enterococcus faecium'\n"
   echo -e "Input:\n \t -i \t Mandatory: Path to the graph file in *.gfa format used to extract nodes and links. Gfa file format\n"
   echo -e "Classifier:\n \t -c \t Mandatory: Classifier used to predict the contigs extracted from the input graph. String value: 'plasflow' or 'mlplasmids'\n"
   echo -e "Bacterial species: \n \t -s \t Mandatory (if mlplasmids is chosen): Bacterial species from the graph file. If you have specified mlplasmids as classifier
                 you need to provide one of the following three bacterial species:\n
                'Enterococcus faecium','Klebsiella pneumoniae' or 'Escherichia coli'\n"

   echo -e "Output name:\n \t -n \t Optional: Output name used in the files generated by gplas. Default: 'unnamed'\n"
   echo -e "Settings:\n \t -t \t Optional: Threshold to predict plasmid-derived sequences. Integer value ranging from 0 to 1.
                 Default mlplasmids threshold: 0.5
                 Default plasflow threshold: 0.7\n"
   echo -e "\t -x \t Optional: Number of times gplas finds plasmid paths per each plasmid starting node. Integer value ranging from 1 to infinite.
                 Default: 20\n"
   echo -e "\t -f \t Optional: Gplas filtering threshold score to reject possible outcoming edges. Integer value ranging from 0 to 1.
                 Default: 0.1\n"
   echo -e "Benchmarking purposes: \n \t -r \t Optional: Path to the complete reference genome corresponding to the graph given. Fasta file format"
   exit
   ;;
   i)
     input=$OPTARG
     ;;
   c)
     classifier=$OPTARG
     ;;
   n)
     name=$OPTARG
     ;;
   s)
     species=$OPTARG
     ;;
   h)
     help=$OPTARG
     ;;
   t)
     threshold_prediction=$OPTARG
     ;;
   x)
     number_iterations=$OPTARG
     ;;
   f)
     filt_gplas=$OPTARG
     ;;
   r)
     reference=$OPTARG
     ;;
   \?)
     ./gplas.sh -h
     echo -e "\n"
     echo "Invalid option: -$OPTARG" >&2
     echo "gplas requires a single input corresponding to a gfa file, use the following syntax:"
     exit 1
     ;;
   :)
     ./gplas.sh -h
     echo -e "\n"
     echo "Error: Option -$OPTARG requires an argument." >&2
     exit 1
     ;;
 esac
done

if [ -z "$input" ];
then
    ./gplas.sh -h
    echo -e "\n Error: Ups, it seems that you are missing the input graph.\n"
    exit
fi

if [ -z "$classifier" ];
then
    ./gplas.sh -h
    echo -e "\n Error: Ups, it seems that you are missing the classifier that you want to use.\n"
    exit
fi

if [ "$classifier" != "plasflow" ];
then
  if [ "$classifier" != "mlplasmids" ];
  then
    ./gplas.sh -h
    echo -e "\n Error: Ups, it seems that you have given a classifier not listed in gplas\n"
    exit
  fi
fi



if [ "$classifier" == "mlplasmids" ];
then
  if [ -z "$species" ];
  then
      ./gplas.sh -h
      echo -e "\n Error: You have specified mlplasmids as classifier you have not indicidate one of the following three bacterial species:
      \t'Enterococcus faecium','Klebsiella pneumoniae' or 'Escherichia coli'\n"
      exit
  fi
fi



if [ "$classifier" == "mlplasmids" ];
then
  if [ -z "$species" ];
   then
    ./gplas.sh -h
    echo -e "\n Error: Remember to indicate one of the following three bacterial species: 'Enterococcus faecium','Klebsiella pneumoniae' or 'Escherichia coli'.\n"
    exit
   fi
fi

echo -e "\n"

cat figures/logo.txt

echo -e "\n"
echo "##################################################################"
echo -e "\n"

echo -e "This is your input graph:" $input "\n"

if [ -z "$species" ];
then
    echo -e "You have not indicated the bacterial species that you are using in the graph\n"
    species="unknown"
else
  echo -e "This is the bacterial species that you have indicated:" $species "\n"
  species=$(echo "'"$species"'")
fi

if [ -z "$name" ];
then
    echo -e "You have not passed an output name. Your results will be named as 'unnamed''\n"
    name="unnamed"
else
  echo -e "Your results will be named using" $name "\n"
fi

if [ -z "$threshold_prediction" ];
then
  if [ "$classifier" == "plasflow" ];
  then
    echo -e "You have not indicated a threshold prediction. Using 0.7 since you are using plasflow\n"
    threshold_prediction=0.7
  else
    echo -e "You have not indicated a threshold prediction. Using 0.5 since you are using mlplasmids\n"
    threshold_prediction=0.5
  fi
else
  echo -e "You have indicated a threshold prediction of:" $threshold_prediction "\n"
fi

if [ -z "$filt_gplas" ];
then
    filt_gplas=0.1
else
    echo -e "Using the following gplas filtering threshold score:" $filt_gplas "\n"
fi

if [ -z "$number_iterations" ];
then
    echo -e "You have not passed the number of times to look for paths based on each plasmid seed, using 20 as default\n"
    number_iterations=20
else
  echo -e "You have indicated a number of iterations of:" $number_iterations "\n"
fi

if [ -z "$reference" ];
then
    reference="No reference provided"
else
  echo -e "You have given a reference genome to evaluate the results of gplas:" $reference
  mkdir -p reference_genome
  rm reference_genome/*
  cp $reference reference_genome/
  mv reference_genome/*.fasta reference_genome/"$name"_ref_genome.fasta
fi


echo "##################################################################"

( echo "cat <<EOF >final.yaml";
  cat template.yml;
  echo "EOF";
) >temp.yaml
. temp.yaml

sleep 10s

if command -v conda > /dev/null; then
 echo  -e 'Conda is present\n'
else
 echo -e "We need conda to run gplas.\n Installing conda"
 wget https://repo.continuum.io/miniconda/Miniconda3-latest-Linux-x86_64.sh
 chmod +x Miniconda3-latest-Linux-x86_64.sh
 mkdir -p ~/tmp
 ./Miniconda3-latest-Linux-x86_64.sh -b -p ~/tmp/Miniconda3
 rm Miniconda3-latest-Linux-x86_64.sh
 export PATH=~/tmp/Miniconda3/bin:$PATH
 export PYTHONPATH=~/tmp/Miniconda3/pkgs/
 conda config --add channels defaults
 conda config --add channels bioconda
 conda config --add channels conda-forge
 export PERL5LIB=~/tmp/Miniconda3/lib/perl5/site_perl/5.22.0
fi


echo -e "Creating a conda environment to install and run snakemake"
source activate gplas || conda create --name gplas --file spec-snakemake.txt
source activate gplas


if [ "$classifier" == "mlplasmids" ];
then
  if [ "$reference" == "No reference provided" ];
  then
      snakemake --unlock --use-conda -s mlplasmidssnake.smk results/"$name"_results.tab
      snakemake --use-conda -s mlplasmidssnake.smk results/"$name"_results.tab
  else
      snakemake --unlock --use-conda -s mlplasmidssnake.smk evaluation/"$name"_completeness.tab
      snakemake --use-conda -s mlplasmidssnake.smk evaluation/"$name"_completeness.tab
  fi
else
  if [ "$reference" == "No reference provided" ];
  then
      snakemake --unlock --use-conda -s plasflowsnake.smk results/"$name"_results.tab
      snakemake --use-conda -s plasflowsnake.smk results/"$name"_results.tab
  else
      snakemake --unlock --use-conda -s plasflowsnake.smk evaluation/"$name"_completeness.tab
      snakemake --use-conda -s plasflowsnake.smk evaluation/"$name"_completeness.tab
  fi
fi

file_to_check=results/"$name"_results.tab

if [ -f "$file_to_check" ];
then
  cat figures/logo.txt
  echo -e "\n"
  echo -e "Congratulations! Prediction succesfully done.\n"
  echo -e "Input graph:" $input "\n"
  echo -e "Bacterial species: " $species "\n"
  echo -e "Classifier:" $classifier "\n"
  echo -e "Threshold for predicting plasmid-derived contigs: $threshold_prediction\n"
  echo -e "Number of plasmid paths created per node: $number_iterations\n"
  echo -e "Threshold of gplas scores: $filt_gplas\n"
  echo -e "\n"

  echo -e "Your results are present at results/ and path/\n"
  echo -e "We hope it helps in your research, thanks for using gplas\n"
  echo -e "If you have used plasflow as a classifier please cite:
  Pawel S Krawczyk et al. PlasFlow: predicting plasmid sequences in metagenomic data using genome signatures, Nucleic Acids Research, doi: 10.1093/nar/gkx1321"
  echo -e "\n"
  echo -e "If you have used mlplasmids as a classifier please cite:
  Arredondo-Alonso et al. mlplasmids: a user-friendly tool to predict plasmid- and chromosome-derived sequences for single species, Microbial Genomics, doi: 10.1099/mgen.0.000224"
  echo -e "\n"

  echo -e "Preprint of gplas coming soon, hasta la vista!"
else
  echo -e "Seems like something went wrong!"
fi
