---
title: "gplas: binning plasmid-predicted contigs"
output: 
  github_document
fontsize: 16pt
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Introduction 

gplas is a tool to bin plasmid-predicted contigs using co-occurence networks. Gplas is a new implementation of mlplasmids which adds the possibility of accurately binning predicted plasmid contigs into several independent components. We rely on the structure of the assembly graph and k-mer information to create a new plasmidome network and predict plasmid contigs belonging to different genomic units.

```{r, echo = FALSE}
knitr::include_graphics("figures/logo.png")
```

# Installation 

```{bash, eval = FALSE}
git clone https://gitlab.com/sirarredondo/gplas.git
cd gplas
./gplas.sh -i test/faecium_graph.gfa
```

First-time installation can take some time depending on your internet speed. But! You do not have to worry about installing any dependencies or running the pipeline by yourself. Thanks to conda and snakemake, we integrated all the dependencies required to run gplas in different conda environments depending on the part of the pipeline. 

After the first-time installation, you will get the prediction of gplas in a few minutes and using a single thread. 


Gplas will check if the following tools are present in your system:

1. [Conda](https://bioconda.github.io/)

2. [Snakemake](https://snakemake.readthedocs.io/en/stable/)

After this, gplas will start the snakemake pipeline and will install different conda environments with the following R packages:

  &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; [igraph](https://cran.r-project.org/web/packages/igraph/index.html) version 1.2.4.1
  
  &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; [ggraph](https://cran.r-project.org/web/packages/ggraph/index.html) version 1.0.2
  
  &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; [Biostrings](https://www.bioconductor.org/packages/release/bioc/html/Biostrings.html) version 2.50.2
  
  &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; [seqinr](https://cran.r-project.org/web/packages/seqinr/index.html) version 3.4-5
  
  &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; [tidyverse](https://www.tidyverse.org/) version 1.2.1 
  
  &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; [spatstat](https://cran.r-project.org/web/packages/spatstat/index.html) version 1.59-0
  
  &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; [cooccur](https://cran.r-project.org/web/packages/cooccur/index.html) version 1.3
  
  &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; [ggrepel](https://cran.r-project.org/web/packages/ggrepel/index.html) version 0.8.0 

Following this, it will install the tools that we use to predict plasmid-derived contigs. 

4. [mlplasmids](https://gitlab.com/sirarredondo/mlplasmids) version 1.0.0 

5. [plasflow](https://anaconda.org/bioconda/plasflow) version 1.1


# Usage 

## Quick usage 

### Running gplas just providing the assembly graph

Gplas only requires a single argument *'-i'* corresponding to an assembly graph in gfa format. 


```{r, echo = FALSE}
knitr::include_graphics("figures/graph.png")
```


```{bash, eval = FALSE}
./gplas.sh -i test/faecium_graph.gfa
```


## Complete usage 

Gplas can take the following arguments:

Mandatory arguments: 

- **-i**: Path to the graph file in *.gfa format used to extract nodes and links. Gfa file format

Optional arguments:

- **-n**: Project name given to gplas. Default: 'unnamed'
- **-s**: Bacterial species from the graph file. If bacterial species corresponds to: 'Enterococcus faecium','Klebsiella pneumoniae' or 'Escherichia coli' then prediction will be perfomed using mlplasmids. Default: 'unknown'
- **-t**: Threshold to predict plasmid-derived sequences. Integer value ranging from 0 to 1. Default: 0.5
- **-x**: Number of times gplas finds plasmid paths per each plasmid starting node. Integer value ranging from 1 to infinite. Default: 10
- **-m**: Mode to run gplas: 'normal' or 'bold'. Bold mode increases the acceptance of connections to enlogate the path. String value. Default: 'normal'

For benchmarking purposes you can pass a complete genome to gplas and will generate a precision and completeness. Using this you can assess the performance of gplas on a small set of genomes in which perhaps you have generated long-reads. 

- **-r**: Path to the complete reference genome corresponding to the graph given. Fasta file format


For example, if we want to use mlplasmids with the *Enterococcus faecium* model on the same test set we only need to specify the following:

```{r, echo = FALSE}
knitr::include_graphics("figures/mlplasmids_logo.png")
```


```{bash, eval = FALSE}
./gplas.sh -i test/faecium_graph.gfa -n 'using_mlplasmids' -s 'Enterococcus faecium'
```

# Help page 


```{bash}
./gplas.sh -h
```


# Main output files 

Gplas will create a folder called 'results' with the following files: 

```{bash}
ls results/using_mlplasmids*
```

## results/

### results/*results.tab 

Tab delimited file containing the prediction given by mlplasmids or plasflow together with the bin prediction assigned by gplas. The file contains the following information: contig number, probability of being chromosome-derived, probability of being plasmid-derived,  class prediction, contig name, k-mer coverage, length, component assigned. 

```{r, echo = FALSE}
results <- read.table(file = 'results/using_mlplasmids_results.tab', header = TRUE)
knitr::kable(results)
```

### results/*components.tab

Tab delimited file containing the bin prediction reported by gplas with the following information: contig number, component assignation  

```{r, echo = FALSE}
components <- read.table(file = 'results/using_mlplasmids_components.tab', header = TRUE)

knitr::kable(components)
```

### results/*plasmidome_network.png

Png file of the plasmidome network generated by gplas after creating an undirected graph using the significant co-occurrence links corresponding to plasmid starting nodes. 

```{r, echo=FALSE}
knitr::include_graphics("results/using_mlplasmids_plasmidome_network.png")
```

### results/*components.fasta

Fasta files with the nodes belonging to each predicted component. 

```{bash}
grep '>' results/using_mlplasmids*.fasta
```

## paths/ 

### paths/*solutions.csv

gplas generates plasmid-like paths per each plasmid starting node. These paths are used later on to construct the co-occurrence networks but they can also be useful to observe all the different paths starting from a single node. These paths can be directly given to Bandage to visualize and manually inspect a path. 

In this case, we find different possible plasmid paths starting from the node 18+. These paths may contain inversions and rearrangements since repeats units such as transposases can be present several times in the same plasmid sequence. In these cases, gplas can traverse the sequence in different ways generating different plasmid-like paths. 

```{bash}
head -n 10 paths/using_mlplasmids_solutions.csv
```

For example, we can inspect in Bandage the path: 18+,76-,52+,94+,57-,77+,18+ 

This path forms a circular sequence since there is overlap between the initial and end node of the path. 

```{r, echo=FALSE}
knitr::include_graphics("figures/bandage_path.jpg")
```

# Issues/Bugs 

You can report any issues or bugs that you find while installing/running gplas using the issue tracker

