---
title: "gplas documentation"
output: 
  github_document
fontsize: 16pt
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Gplas

gplas is a tool to bin plasmid-predicted contigs using co-occurence networks. Gplas is a new implementation of mlplasmids which adds the possibility of accurately binning predicted plasmid contigs into several independent components. We rely on the structure of the assembly graph and k-mer information to create a new plasmidome network and predict plasmid contigs belonging to different genomic units.

```{r, echo = FALSE}
knitr::include_graphics("figures/logo.png")
```

## Installation 

```{bash, eval = FALSE}
git clone https://gitlab.com/sirarredondo/gplas.git
cd gplas
```

### Installation of all the dependencies using a test graph

```{bash, eval = FALSE}
./gplas.sh -i test/faecium_graph.gfa
```

First-time installation can take some time (you don't have to do anything at all, so go for a coffee and come back in 10m). gplas needs to check whether:

1. **Conda** is installed. 

2. **Snakemake** is present in the conda environment. 

3. Install all the **R libraries** required to run gplas.

  &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;**igraph** version 1.2.4.1
  
  &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;**ggraph** version 1.0.2
  
  &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;**Biostrings** version 2.50.2
  
  &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;**seqinr** version 3.4-5
  
  &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;**tidyverse** version 1.2.1 
  
  &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;**spatstat** version 1.59-0
  
  &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;**cooccur** version 1.3
  
  &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;**ggrepel** version 0.8.0 


4. **mlplasmids** (version 1.0.0) is present in the conda environment  

5. **Plasflow** (version 1.1) is present in the conda environment. 


## Usage 

### Quick usage 

**Running gplas providing uniquely the assembly graph**

Gplas only requires a single argument '-i' corresponding to an assembly graph in gfa format. 

```{bash, eval = FALSE}
./gplas.sh -i test/faecium_graph.gfa
```

### Complete usage 

Gplas can take the following arguments:

Mandatory arguments: 

- **-i**: Path to the graph file in *.gfa format used to extract nodes and links. Gfa file format

Optional arguments:

- **-n**: Project name given to gplas. Default: 'unnamed'
- **-s**: Bacterial species from the graph file. If bacterial species corresponds to: 'Enterococcus faecium','Klebsiella pneumoniae' or 'Escherichia coli' then prediction will be perfomed using mlplasmids. Default: 'unknown'
- **-t**: Threshold to predict plasmid-derived sequences. Integer value ranging from 0 to 1. Default: 0.5
- **-x**: Number of times gplas finds plasmid paths per each plasmid starting node. Integer value ranging from 1 to infinite. Default: 10
- **-m**: Mode to run gplas: 'normal' or 'bold'. Bold mode increases the acceptance of connections to enlogate the path. String value. Default: 'normal'

For benchmarking purposes you can pass a complete genome to gplas and will generate a precision and completeness. Using this you can assess the performance of gplas on a small set of genomes in which perhaps you have generated long-reads. 

- **r**: Path to the complete reference genome corresponding to the graph given. Fasta file format


For example, if we want to use mlplasmids with the *Enterococcus faecium* model on the same test set we only need to specify the following:

```{bash, eval = FALSE}
./gplas.sh -i test/faecium_graph.gfa -n 'using_mlplasmids' -s 'Enterococcus_faecium'
```

## Main output files 

Gplas will create a folder called 'results' with the following files: 

```{bash}
ls results/using_mlplasmids*
```

### results.tab 

Tab delimited file containing the prediction given by mlplasmids or plasflow together with the bin prediction assigned by gplas. The file contains the following information: contig number, probability of being chromosome-derived, probability of being plasmid-derived,  class prediction, contig name, k-mer coverage, length, component assigned. 

```{R}
results <- read.table(file = 'results/using_mlplasmids_results.tab')
results
```

### components.tab

Tab delimited file containing the bin prediction reported by gplas with the following information: contig number, component assignation  

```{r}
components <- read.table(file = 'results/using_mlplasmids_components.tab')
components
```

### plasmidome_network.png

Png file of the plasmidome network generated by gplas after creating an undirected graph using the significant co-occurrence links corresponding to plasmid starting nodes. 

```{r, echo=FALSE}
knitr::include_graphics("results/using_mlplasmids_plasmidome_network.png")
```

### components.fasta

Fasta files with the nodes belonging to each predicted component. 

```{bash}
grep '>' results/unnamed_project*.fasta
```


